README.md
# clappr-detach-plugin

Note: this plugin was previously called `clappr-detach-plugin`, but has changed to `@globocom/clappr-detach-plugin`.

A [Clappr](https://github.com/clappr/clappr) plugin to detach the video.

![clappr-detach-plugin](https://user-images.githubusercontent.com/4842605/30820777-567d2960-a1f9-11e7-854a-32acc6cb0a47.png)

This project is a fork from https://github.com/team-767/clappr-detach-plugin. We took a different approach that was done in that project.


## Installation

```shell
npm install -S @globocom/clappr-detach-plugin
```

## Usage

To init the plugin you have to pass it the Clappr object. See bellow the compatibility between this plugin and Clappr versions.

```
import Clappr from 'clappr'
import initClapprDetachPlugin from '@globocom/clappr-detach-plugin'

const detachPlugin = initClapprDetachPlugin(Clappr)
```

We did it this way so the plugin will receive the dependency and doesn't have to install it.

Full example:

```javascript
import Clappr from 'clappr'
import initClapprDetachPlugin from '@globocom/clappr-detach-plugin'

const player = new Clappr.Player({
  source: 'http://www.sample-videos.com/video/mp4/480/big_buck_bunny_480p_5mb.mp4',
  plugins: {
    'core': [initClapprDetachPlugin(Clappr)]
  }
})
```

This plugin exposes the public methods `attach` and `detach` that you can call with:

```javascript
player.getPlugin('detach').detach()
player.getPlugin('detach').attach()
```

Or using the player `configure` method:

```javascript
player.configure({ isDetached: true })
player.configure({ isDetached: false })
```


## Options

```javascript
import Clappr from 'clappr'
import initClapprDetachPlugin from '@globocom/clappr-detach-plugin'

const player = new Clappr.Player({
  source: 'http://www.sample-videos.com/video/mp4/480/big_buck_bunny_480p_5mb.mp4',
  plugins: {
    'core': [initClapprDetachPlugin(Clappr)]
  },
  // these are the default values
  detachOptions: {
    dragEnabled: true,
    height: 180,
    isDetached: false,
    onAttach: () => { },
    onDetach: () => { }
    opacity: 1,
    orientation: 'bottom-right',
    position: {
      bottom: 10,
      right: 10,
    },
    width: 320,
  }
})
```

| Property            | Type          | Description                                             | valid values                                                          |
| ------------------- | ------------- | ------------------------------------------------------- | --------------------------------------------------------------------- |
| dragEnabled         | bool          | wheter drag n drop of detached player should be enabled | -                                                                     |
| height              | number        | player height on detach                                 | -                                                                     |
| isDetached          | bool          | whether you want to start the player detached           | -                                                                     |
| onAttach            | function      | callback called when player attach on original place    | -                                                                     |
| onDetach            | function      | callback called when player detach from original place  | -                                                                     |
| orientation         | string        | Where the player will appear on detach                  | 'bottom-left', 'top-left', 'bottom-right', 'top-right'                |
| position            | object        | Force the position where player will appear on detach   | `{"top": number, "right": number, "bottom": number, "left": number}`  |
| width               | number        | player width on detach                                  | -                                                                     |


## Compatibility

This table **doesn't** mean that other versions won't work together.

It **does** means that we are sure that these versions work together.

| clappr-detach-plugin | clappr       |
| -------------------- | ------------ |
| 0.4.0                | 0.2.73       |
| 0.2.0                | 0.2.73       |
| 0.1.0                | 0.2.73       |


CHANGELOG.md
# CHANGELOG

## 0.4.7

- added scope to the package, so the name has changed from `clappr-detach-plugin` to `@globocom/clappr-detach-plugin`

## 0.4.6

- adjust detach/attach of plugin

## 0.4.5

- fix publish. Do not use version `0.4.4`

## 0.4.4

- fix publish. Do not use version `0.4.3`

## 0.4.3

- validate whether core is ready before toggling detach state

## 0.4.2

- prevent mouse from becoming stuck in move style

## 0.4.1

- prevent dragging when mouse over bars (seekbar and sound bar)

## 0.4.0

- adds `dragEnabled` option

## 0.3.0

Don't use this version, we messed up the publishing

## 0.2.0

- change approach of detach, to do it using CSS

## 0.1.0

- remove dependency on Clappr and receives Clappr as parameter
- remove dependency on Zepto

## 0.0.4

- Accept an option `isDetached`, that can be called with `player.configure({ isDetached })`
- Removes option `detachOnStart`

## 0.0.3

- do not use

## 0.0.2

- do not use

## 0.0.1

- First version


src/index.js
import assign from 'lodash.assign'

import setupInteractions from './interactions'
import { addDragArea, removeDragArea } from './drag'

// assets
import detachIcon from './assets/detach-icon.svg'
import detachPlaceholder from './assets/detach-placeholder.html'
import detachToggle from './assets/detach-toggle.html'
import detachStyle from './assets/detach.css'

let dragInteractable

const NOOP = () => {}

const DEFAULT_POSITION = 10

const DEFAULT_POSTER = 'data:image/gif;base64,R0lGODlhAQABAIAAAAUEBAAAACwAAAAAAQABAAACAkQBADs='

const DEFAULT_OPTIONS = {
  dragEnabled: true,
  isDetached: false,
  orientation: 'bottom-right',
  position: {
    bottom: DEFAULT_POSITION,
    right: DEFAULT_POSITION,
  },
  opacity: 1,
  width: 320,
  height: 180,
  onAttach: NOOP,
  onDetach: NOOP,
}

const orientationOptions = (orientation, position) => {
  const options = {}

  orientation.split('-').forEach((side) => {
    if (side === 'left') { options.left = position.left || DEFAULT_POSITION }
    if (side === 'right') { options.right = position.right || DEFAULT_POSITION }
    if (side === 'bottom') { options.bottom = position.bottom || DEFAULT_POSITION }
    if (side === 'top') { options.top = position.top || DEFAULT_POSITION }
  })

  return options
}

const initPlugin = ({
  UICorePlugin,
  Events,
  Styler,
  template,
}) => {
  const DETACH_STYLE_TAG = Styler.getStyleFor(detachStyle)

  const DETACH_ICON_SVG = template(detachIcon)()

  const DETACH_TOGGLE_HTML = template(detachToggle)({
    icon: DETACH_ICON_SVG,
  })

  const placeholderMarkup = (poster = DEFAULT_POSTER) => template(detachPlaceholder)({
    icon: DETACH_ICON_SVG,
    backgroundImage: poster,
  })

  return class ClapprDetachPlugin extends UICorePlugin {
    /* eslint-disable class-methods-use-this */
    // clappr uses this - to find the plugin by name
    get name() {
      return 'detach'
    }

    // clappr uses this - to add this class to the player's root element
    get className() {
      return 'clappr-detach'
    }
    /* eslint-enable class-methods-use-this */

    // clappr uses this - to expose like player.getPlugin('detach').detach()
    getExternalInterface() {
      return {
        attach: this.attach,
        detach: this.detach,
      }
    }

    get currentContainer() { return this.core.getCurrentContainer() }

    constructor(core) {
      super(core)

      if (core.ready) {
        this.onCoreReady()
      }
    }

    /*
      ---------------------------------------------------------------------------
      options
      ---------------------------------------------------------------------------
    */
    getOptions() {
      return {
        ...DEFAULT_OPTIONS,
        ...this.core.options.detachOptions,
      }
    }

    setOptions(options) {
      assign(this.core.options.detachOptions, options)
    }

    // when this function is called, `this.core.options.detachOptions` was already changed by clappr,
    // so we can use the new value
    // we have to verify the core first to not apply changes before the right moment
    onOptionsChange() {
      if (this.core.ready) { this.toggleDetach(this.core.options.isDetached) }
    }

    /*
      ---------------------------------------------------------------------------
      elements
      ---------------------------------------------------------------------------
    */
    /*
      initialize references to some elements that later we will use a lot
    */
    initElements() {
      this.initPlayerContainerPlaceholderElement()
    }

    /*
      (jQuery element) $player is the player element itself. We don't instantiate $player, Clappr does
    */
    get $player() { return this.core.$el }

    /*
      (jQuery element) $playerContainer is the element where the $player is attached to when created
    */
    get $playerContainer() { return this.$player.parent() }

    /*
      (jQuery element) $playerPlaceholder is the element we put a placeholder in when we detach the player
    */
    get $playerPlaceholder() { return this.$el }

    initPlayerContainerPlaceholderElement() {
      this.$playerPlaceholder.empty()
      this.$playerPlaceholder.append(placeholderMarkup(this.core.options.poster))
      this.$playerPlaceholder.append(DETACH_STYLE_TAG)
    }

    /*
      ---------------------------------------------------------------------------
      events
      ---------------------------------------------------------------------------
    */
    bindEvents() {
      this.listenTo(this.core, Events.CORE_OPTIONS_CHANGE, this.onOptionsChange)
      this.listenTo(this.core, Events.CORE_READY, this.onCoreReady)
      this.listenTo(this.core, Events.CORE_FULLSCREEN, this.onCoreFullScreen)
      this.listenTo(this.mediaControl, Events.MEDIACONTROL_RENDERED, this.onMediaControlRendered)
    }

    onCoreReady() {
      this.initElements()

      // clean up any element that might conflict with the placeholder
      this.$playerContainer.find(this.className).remove()
      const elementDetach = this.$playerPlaceholder.detach()
      elementDetach.appendTo(this.$playerContainer)

      // save player original styles to reset back to it when needed
      this.playerOriginalStyle = this.$player.attr('style')

      // set the playerPlaceholder styles based on the player styles
      this.$playerPlaceholder.attr('style', this.playerOriginalStyle)

      if (this.getOptions().isDetached) {
        this.toggleDetach(true)
      }
    }

    /*
      ---------------------------------------------------------------------------
      fullscreen
      ---------------------------------------------------------------------------
    */
    onCoreFullScreen(goingToFullScreen) {
      this.mediaControlDetachToggle.toggle(!goingToFullScreen)

      if (goingToFullScreen) {
        return
      }

      // the timeout is needed to delay the execution until the fullscreen has finished quiting
      setTimeout(() => {
        this.updatePlayer(this.getOptions().isDetached)
      })
    }

    /*
      ---------------------------------------------------------------------------
      media control
      ---------------------------------------------------------------------------
    */
    get clickToPausePlugin() { return this.core.containers[0].getPlugin('click_to_pause') }
    get mediaControl() { return this.core.mediaControl }
    get mediaControlDetachToggle() { return this.mediaControl.$el.find('[data-js="clappr-detach__detach-toggle"]') }
    get mediaControlRightPanel() { return this.mediaControl.$el.find('.media-control-right-panel') }
    get mediaControlSeekBar() { return this.mediaControl.$el.find('.media-control-center-panel') }

    /*
      adds the toggle detach button to the media control
    */
    onMediaControlRendered() {
      if (this.mediaControl.setKeepVisible) {
        this.mediaControl.setKeepVisible(true)
      }
      this.mediaControlRightPanel.append(DETACH_TOGGLE_HTML)
      this.mediaControlDetachToggle.on('click', () => {
        this.toggleDetach(!this.getOptions().isDetached)
      })
    }

    /*
      ---------------------------------------------------------------------------
      player
      ---------------------------------------------------------------------------
    */
    updatePlayer(isDetached) {
      this.mediaControlSeekBar.toggle(!isDetached)
      this.$playerPlaceholder.toggleClass('clappr-detach--visible', isDetached)
      this.$player.toggleClass('clappr-detach--is-detached', isDetached)
      this.$player[0].style.transform = 'translate(0, 0)'

      const options = this.detachedOptions
      if (isDetached) {
        this.$player[0].style.opacity = `${options.opacity}`
        this.$player[0].style.left = `${options.left}px`
        this.$player[0].style.right = `${options.right}px`
        this.$player[0].style.top = `${options.top}px`
        this.$player[0].style.bottom = `${options.bottom}px`
        this.$player[0].style.setProperty('width', `${options.width}px`, 'important')
        this.$player[0].style.setProperty('height', `${options.height}px`, 'important')

        if (options.dragEnabled) {
          const result = setupInteractions(this.$player[0], {
            drag: {
              dragArea: addDragArea(),
            },
            drop: {
              dropAreaElement: this.$playerPlaceholder[0],
              onDrop: this.attach,
            },
          })

          dragInteractable = result.drag
        }
      } else {
        this.$player.attr('style', this.playerOriginalStyle)
        if (options.dragEnabled) {
          removeDragArea()
          if (dragInteractable) {
            dragInteractable.unset()
          }
        }
      }
    }

    /*
      ---------------------------------------------------------------------------
      detached
      ---------------------------------------------------------------------------
    */
    get detachedOptions() {
      const { orientation, position, ...options } = this.getOptions()
      return { ...options, ...orientationOptions(orientation, position) }
    }

    /*
      ---------------------------------------------------------------------------
      attach / detach
      ---------------------------------------------------------------------------
    */
    toggleDetach = (isDetached) => {
      this.setOptions({
        isDetached,
      })
      const isPlaying = this.currentContainer.isPlaying()

      this.updatePlayer(isDetached)

      if (isDetached) {
        this.clickToPausePlugin.disable()
        this.getOptions().onDetach()
      } else {
        this.clickToPausePlugin.enable()
        this.getOptions().onAttach()
      }

      // when the player is moved while playing, it becomes paused. So we restore the playing
      if (isPlaying) {
        this.currentContainer.play()
      }
    }

    attach = () => {
      if (!this.getOptions().isDetached) {
        return
      }
      this.toggleDetach(false)
    }

    detach = () => {
      if (this.getOptions().isDetached) {
        return
      }
      this.toggleDetach(true)
    }
  }
}

export default initPlugin
